write-fonts/src/tables/cmap.rs

```diff
@@ -237,6 +236,187 @@ impl Cmap {
         ))
     }
 }
+// a helper for iterating over ranges for cmap 4
+struct Format4Ranges<'a> {
+    mappings: &'a [(char, GlyphId)],
+    seg_start: usize,
+    currently_contiguous: bool,
+}
+
+#[derive(Clone, Copy, Debug)]
+struct Format4Segment {
+    // indices are into the source mappings
+    start_ix: usize,
+    end_ix: usize,
+    id_delta: Option<i32>,
+}
+
+impl Format4Segment {
+    fn len(&self) -> usize {
+        self.end_ix - self.start_ix + 1
+    }
+
+    // cost in bytes of this segment
+    fn cost(&self) -> usize {
+        const BASE_COST: usize = 8; // 4 u16s for a new segment
+        let glyph_id_cost = self
+            .id_delta
+            .is_none()
+            .then(|| self.len() * u16::RAW_BYTE_LEN)
+            .unwrap_or(0);
+        BASE_COST + glyph_id_cost
+    }
+
+    fn combine(&self, other: Format4Segment) -> Format4Segment {
+        assert_eq!(other.start_ix, self.end_ix + 1);
+        Format4Segment {
+            start_ix: self.start_ix,
+            end_ix: other.end_ix,
+            id_delta: None,
+        }
+    }
+}
+
+impl<'a> Format4Ranges<'a> {
+    fn new(mappings: &'a [(char, GlyphId)]) -> Self {
+        // ignore chars above BMP:
+        let mappings = mappings
+            .iter()
+            .position(|(c, _)| u16::try_from(*c as u32).is_err())
+            .map(|bad_idx| &mappings[..bad_idx])
+            .unwrap_or(mappings);
+        Self {
+            mappings,
+            seg_start: 0,
+            currently_contiguous: false,
+        }
+    }
+
+    /// a convenience method called from our iter in the various cases where
+    /// we emit a segment.
+    ///
+    /// a 'seg_len' of 0 means start == end, e.g. a segment of one glyph.
+    fn make_segment(&mut self, seg_len: usize) -> Format4Segment {
+        // if start == end, we should always use a delta.
+        let use_delta = self.currently_contiguous || seg_len == 0;
+        let result = Format4Segment {
+            start_ix: self.seg_start,
+            end_ix: self.seg_start + seg_len,
+            id_delta: use_delta
+                .then_some(
+                    self.mappings
+                        .get(self.seg_start)
+                        .map(|(cp, gid)| gid.to_u32() as i32 - *cp as u32 as i32),
+                )
+                .flatten(),
+        };
+        self.seg_start += seg_len + 1;
+        self.currently_contiguous = false;
+        result
+    }
+}
+
+// this iterator creates segments greedily, that is it will create a new segment
+// at every opportunity. These are then recombined by the caller to generate
+// the most efficient overall sequence of segments.
```

Comment:
I'm not quite sure what this is trying to tell me, why this is important. Perhaps we could mention that?

Reply:
Actually that sort of describes the code in general. I think it's saying it tries to find runs, prioritizing, from most to least favorable:

1. A run of at least 2 where codepoint *and* gid are contiguous
1. A run of at least 2 where codepoints are contiguous (but gids aren't)
1. A run of 1

I think the comment might want to elaborate.

Reply:
https://github.com/googlefonts/fontations/pull/1229 attempts to simplify, eliminating the currently_contiguous field and making the iterator just grab the next run, prioritized as ^, each cycle.

Reply:
I'm trying to explain a slightly arbitrary design choice; that this iterator returns segments but that they need further optimization. In my initial design the iterator produced the segments directly, but you need to do some lookahead to determine if you should split a segment or not.